File: promptvault/backend/internal/usecases/audit/audit.go
package audit

import (
	"context"
	"encoding/json"

	repo "promptvault/internal/interface/repository"
	"promptvault/internal/models"
)

// Service — usecase для записи админ-действий в audit_log.
// Все поля admin_id/ip/user_agent автоматически прокидываются через
// AdminRequestInfo из context (см. middleware/admin.AdminAuditContext).
type Service struct {
	repo repo.AuditRepository
}

func NewService(auditRepo repo.AuditRepository) *Service {
	return &Service{repo: auditRepo}
}

// LogInput — минимальный набор параметров для audit.Log.
// BeforeState/AfterState — arbitrary struct/map, будет marshal в JSON.
// Не кладите сюда password_hash, TOTP secrets или JWT — это PII utrust.
type LogInput struct {
	Action      Action
	TargetType  TargetType
	TargetID    *uint
	BeforeState any
	AfterState  any
}

// Log записывает одно событие в audit_log. AdminID/IP/UserAgent берутся из ctx.
// Если ctx не содержит AdminRequestInfo — возвращает ErrMissingRequestInfo.
// Вызывающая сторона должна либо (а) добавить AdminAuditContext middleware
// к роуту, либо (б) явно вручную положить AdminRequestInfo через WithContext.
func (s *Service) Log(ctx context.Context, in LogInput) error {
	info, ok := FromContext(ctx)
	if !ok {
		return ErrMissingRequestInfo
	}

	before, err := marshalState(in.BeforeState)
	if err != nil {
		return err
	}
	after, err := marshalState(in.AfterState)
	if err != nil {
		return err
	}

	entry := &models.AuditLog{
		AdminID:     info.AdminID,
		Action:      string(in.Action),
		TargetType:  string(in.TargetType),
		TargetID:    in.TargetID,
		BeforeState: before,
		AfterState:  after,
		IP:          info.IP,
		UserAgent:   info.UserAgent,
	}
	return s.repo.Log(ctx, entry)
}

// List — pass-through к repo.List для admin audit log feed endpoint.
func (s *Service) List(ctx context.Context, filter repo.AuditLogFilter) ([]models.AuditLog, int64, error) {
	return s.repo.List(ctx, filter)
}

// marshalState конвертит arbitrary state в json.RawMessage для jsonb-колонки.
// nil → nil (не marshal'ить в "null"), иначе — json.Marshal.
func marshalState(state any) ([]byte, error) {
	if state == nil {
		return nil, nil
	}
	return json.Marshal(state)
}

File: promptvault/frontend/src/api/admin/audit.ts
import { api } from "@/api/client"
import type { PaginatedResponse } from "@/api/types"

export interface AuditEntry {
  id: number
  admin_id: number
  action: string
  target_type: string
  target_id?: number
  before_state?: unknown
  after_state?: unknown
  ip: string
  user_agent?: string
  created_at: string
}

export interface AuditFilter {
  action?: string
  target_type?: string
  admin_id?: number
  target_id?: number
  from?: string
  to?: string
  page?: number
  page_size?: number
}

export function fetchAudit(filter: AuditFilter = {}) {
  const params = new URLSearchParams()
  if (filter.action) params.set("action", filter.action)
  if (filter.target_type) params.set("target_type", filter.target_type)
  if (filter.admin_id) params.set("admin_id", String(filter.admin_id))
  if (filter.target_id) params.set("target_id", String(filter.target_id))
  if (filter.from) params.set("from", filter.from)
  if (filter.to) params.set("to", filter.to)
  if (filter.page) params.set("page", String(filter.page))
  if (filter.page_size) params.set("page_size", String(filter.page_size))
  const qs = params.toString()
  return api<PaginatedResponse<AuditEntry>>(`/admin/audit${qs ? `?${qs}` : ""}`)
}
